### PR TITLE
Add timeout handling to on_exit callback in ExUnit

### PR DESCRIPTION
**Pull request by lexmag**

### CODE REVIEW

**Comment by josevalim** [comment]
:shipit:

File: lib/ex_unit/lib/ex_unit/on_exit_handler.ex [hunk 1]
```diff
@@ -54,6 +54,20 @@ defmodule ExUnit.OnExitHandler do
...
+            Process.info(runner_pid, :current_stacktrace)
+          catch
+            _, _ -> []
+          else
+            {:current_stacktrace, stacktrace} -> stacktrace
+          end
+        Process.exit(runner_pid, :kill)
+        Process.demonitor(runner_monitor, [:flush])
+        exception = ExUnit.TimeoutError.exception(timeout: timeout, type: :on_exit)
+        {nil, nil, state || {:error, exception, stacktrace}}
```
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: The stacktrace should be part of the exception and not used as a fake stacktrace, that is very misleading.

File: lib/ex_unit/lib/ex_unit/on_exit_handler.ex [hunk 2]
```diff
@@ -54,6 +54,20 @@ defmodule ExUnit.OnExitHandler do
         {runner_pid, runner_monitor, state || error}
       {:DOWN, ^runner_monitor, :process, ^runner_pid, error} ->
         {nil, nil, state || {{:EXIT, runner_pid}, error, []}}
+    after
+      timeout ->
+        stacktrace =
+          try do
+            Process.info(runner_pid, :current_stacktrace)
+          catch
+            _, _ -> []
```
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: The stacktrace will be `nil` if `runner_pid` doesnt exist and wont raise.

Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: Oh, indeed. That's what you get when you blindly copy code from [another place](https://github.com/elixir-lang/elixir/blob/85d3d8171482ec041f840c25f16f64e85acde5ff/lib/ex_unit/lib/ex_unit/runner.ex#L272-L281). 😄
Will fix that in there as well.

File: lib/ex_unit/lib/ex_unit/on_exit_handler.ex (hunk 1 above)
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: Another PR is coming.

**Pull request merged**